### PR TITLE
Implementation of DummyInferenceEngine

### PR DESCRIPTION
**Pull request by rahat2134**
This PR addresses issue #325  by implementing a DummyInferenceEngine for testing purposes.

## Changes
- Created `DummyInferenceEngine` class in `exo/inference/dummy_inference_engine.py`
- Modified `exo/inference/inference_engine.py` to include the dummy engine option
- Updated `exo/main.py` to support selection of the dummy engine
- Added basic testing for the DummyInferenceEngine

## Features
- Simulates inference without loading a real model
- Generates random token outputs
- Simulates latency using `asyncio.sleep`
- Fully asynchronous implementation
- Simulates shard loading

## Testing
- Added basic test in `main.py` to verify DummyInferenceEngine functionality
- Manually tested with `python main.py --inference-engine dummy`

Feedback and suggestions for improvement are welcome!

### CODE REVIEW

**Comment by rahat2134** [comment]
@AlexCheema PTAL.

File: exo/inference/dummy_inference_engine.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+from typing import Optional, Tuple, TYPE_CHECKING
+import numpy as np
+import asyncio
+import json
+from typing import Optional, Tuple
+if TYPE_CHECKING:
+    from exo.inference.inference_engine import InferenceEngine
```
Comment (AlexCheema) [inline exo/inference/dummy_inference_engine.py:7]: What's the purpose of this? Isn't this redundant?

File: exo/inference/inference_engine.py [hunk 2]
```diff
@@ -17,6 +17,7 @@ async def infer_tensor(self, request_id: str, shard: Shard, input_data: np.ndarr
 
 
 def get_inference_engine(inference_engine_name: str, shard_downloader: 'ShardDownloader'):
+  print(f"get_inference_engine called with: {inference_engine_name}")  # Debug print
```
Comment (AlexCheema) [inline exo/inference/inference_engine.py:20]: if you want a log here, please add `if DEBUG >= 2:`

File: exo/inference/inference_engine.py [hunk 3]
```diff
@@ -27,5 +28,8 @@ def get_inference_engine(inference_engine_name: str, shard_downloader: 'ShardDow
     tinygrad.helpers.DEBUG.value = int(os.getenv("TINYGRAD_DEBUG", default="0"))
 
     return TinygradDynamicShardInferenceEngine(shard_downloader)
+  elif inference_engine_name == "dummy":
+    from exo.inference.dummy_inference_engine import DummyInferenceEngine
+    return DummyInferenceEngine(shard_downloader)
   else:
-    raise ValueError(f"Inference engine {inference_engine_name} not supported")
+      raise ValueError(f"Inference engine {inference_engine_name} not supported. Supported engines are 'mlx', 'tinygrad', and 'dummy'.")
```
Comment (AlexCheema) [inline exo/inference/inference_engine.py:35]: fix indent

File: exo/inference/inference_engine.py [hunk 4]
```diff
@@ -27,5 +28,8 @@ def get_inference_engine(inference_engine_name: str, shard_downloader: 'ShardDow
     tinygrad.helpers.DEBUG.value = int(os.getenv("TINYGRAD_DEBUG", default="0"))
 
     return TinygradDynamicShardInferenceEngine(shard_downloader)
+  elif inference_engine_name == "dummy":
+    from exo.inference.dummy_inference_engine import DummyInferenceEngine
+    return DummyInferenceEngine(shard_downloader)
   else:
```
Comment (AlexCheema) [inline exo/inference/inference_engine.py:34]: remove the else completely. if we get through to the end throw erorr

File: exo/main.py [hunk 5]
```diff
@@ -56,6 +59,15 @@
...
 inference_engine_name = args.inference_engine or ("mlx" if system_info == "Apple Silicon Mac" else "tinygrad")
+print(f"Inference engine name after selection: {inference_engine_name}")
+
+if inference_engine_name not in ["mlx", "tinygrad", "dummy"]:
+  print(f"Warning: Unknown inference engine '{inference_engine_name}'. Defaulting to 'tinygrad'.")
+  inference_engine_name = "tinygrad"
+else:
+  print(f"Using selected inference engine: {inference_engine_name}")
+
+print(f"About to call get_inference_engine with: {inference_engine_name}")
```
Comment (AlexCheema) [inline exo/main.py:70]: No need for any of this. If they choose to override the inference engine name with an unsupported inference engine they'll get an exception which is correct.

File: exo/main.py [hunk 6]
```diff
@@ -173,6 +185,16 @@ async def run_model_cli(node: Node, inference_engine: InferenceEngine, model_nam
...
 
 
+async def test_dummy_inference(inference_engine):
+    print("Testing DummyInferenceEngine...")
+    test_shard = Shard(model_id="test_model", start_layer=0, end_layer=1, n_layers=1)
+    test_prompt = "This is a test prompt"
+    result, state, is_finished = await inference_engine.infer_prompt("test_request", test_shard, test_prompt)
+    print(f"Inference result shape: {result.shape}")
+    print(f"Inference state: {state}")
+    print(f"Is finished: {is_finished}")
```
Comment (AlexCheema) [inline exo/main.py:195]: This shouldn't be in `main.py`

File: exo/main.py [hunk 7]
```diff
@@ -193,6 +215,8 @@ def handle_exit():
     await run_model_cli(node, inference_engine, model_name, args.prompt)
   else:
     asyncio.create_task(api.run(port=args.chatgpt_api_port))  # Start the API server as a non-blocking task
+    if isinstance(node.inference_engine, DummyInferenceEngine):
+      await test_dummy_inference(node.inference_engine)
```
Comment (AlexCheema) [inline exo/main.py:219]: This shouldn't be in `main.py`

File: exo/inference/dummy_inference_engine.py
Comment (AlexCheema) [inline exo/inference/dummy_inference_engine.py:1]: Could you also write an integration test in `.circleci/config.yml` with the dummy inference engine. I think this is a nice test to have to quickly test everything **except** the inference engine itself.

**Review by AlexCheema — CHANGES_REQUESTED** [review]
The implementation of the `DummyInferenceEngine` looks correct. Please make the changes I mentioned and then I will also test on my end end-to-end.

File: exo/inference/dummy_inference_engine.py (hunk 1 above)
Comment (rahat2134) [reply exo/inference/dummy_inference_engine.py]: Done

File: exo/inference/inference_engine.py (hunk 2 above)
Comment (rahat2134) [reply exo/inference/inference_engine.py]: Done Thanks.

File: exo/inference/inference_engine.py (hunk 3 above)
Comment (rahat2134) [reply exo/inference/inference_engine.py]: Done Thanks.

File: exo/inference/inference_engine.py (hunk 4 above)
Comment (rahat2134) [reply exo/inference/inference_engine.py]: Done Thanks.

File: exo/main.py (hunk 5 above)
Comment (rahat2134) [reply exo/main.py]: Done Thanks.

File: exo/main.py (hunk 6 above)
Comment (rahat2134) [reply exo/main.py]: Done Thanks. Added in test file.

File: exo/main.py (hunk 7 above)
Comment (rahat2134) [reply exo/main.py]: Done Thanks. (same as previous)

File: exo/inference/dummy_inference_engine.py
Comment (rahat2134) [reply exo/inference/dummy_inference_engine.py:1]: done thanks

**Comment by rahat2134** [comment]
@AlexCheema PTAL!

**Comment by rahat2134** [comment]
@AlexCheema PTAL.

**Comment by rahat2134** [comment]
@AlexCheema PTAL.

**Comment by AlexCheema** [comment]
Good job, lgtm

**Pull request merged**